feat(ivy): index template elements

ayazhafiz opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
Indexer module does not index the location of element selectors.

Issue Number: #31299

## What is the new behavior?
Index the location of element selectors.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Closes #31299

ayazhafiz commented:
Blocked by #31151. Waiting for merge.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/src/template.ts:
```diff
@@ -142,12 +151,31 @@ class TemplateVisitor extends RecursiveTemplateVisitor {
+    // `<element></element>`. Only the selector is interesting to the indexer, so the source is
+    // searched for the first occurence of the element (selector) name.
+    const start = sourceSpan.start.offset + sourceSpan.toString().indexOf(name);
+    this.identifiers.add({
```
Where are we going to convey the information that the element matches some component/directive(s)?

ayazhafiz replied:
I think that is the job of the actual indexer, not the indexing module inside the compiler. This is especially because every selector is indexed, not the ones that refer to external components.

ayazhafiz replied:
Indexed elements now also include their used directives. Marking as resolved.

ayazhafiz commented on packages/compiler-cli/src/ngtsc/indexer/src/template.ts:
```diff
@@ -7,9 +7,12 @@
  */
 
 import {AST, BoundTarget, DirectiveMeta, ImplicitReceiver, MethodCall, PropertyRead, RecursiveAstVisitor} from '@angular/compiler';
+import {ParseSourceSpan} from '@angular/compiler/src/compiler';
```
Done.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/src/api.ts:
```diff
@@ -48,3 +40,43 @@ export interface IndexedComponent {
+ */
+export interface AttributeIdentifier extends TemplateIdentifier {
+  kind: IdentifierKind.Attribute;
+  value: string;
```
Do we need the `value` of attributes anywhere?

ayazhafiz replied:
No, I'll remove it.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/src/api.ts:
```diff
@@ -48,3 +40,43 @@ export interface IndexedComponent {
+export interface ElementIdentifier extends TemplateIdentifier {
+  kind: IdentifierKind.Element;
+  attributes: AttributeIdentifier[];
+  usedDirectives: ts.Declaration[];
```
The API is starting to look a little strange. Maybe we want to introduce an `Element` type instead? Otherwise it's unclear whether `AttributeIdentifier`s live on the `ElementIdentifier` or on the `IndexedComponent.template.identifiers`.

Also, now there's overlap between `IndexedComponent.template.usedComponents` and `usedDirectives` here. These are also strange fields to have on an `Identifier` which should really only be giving information about the text of the tag name, not the whole element.

ayazhafiz replied:
I agree that the overlap may be a little strange, but it is also appropriate to put information about `usedDirectives` on the element because we have to give information about directives applied to it (the indexer can't determine that without reimplementing `BoundTarget`).

I've been modeling `TemplateIdentifier`s after how TypeScript does `ts.Node`s - there is some base information, like the span and kind, but specialized nodes contain a lot more extra information, like children, arguments, etc. We introduce an `Element` type as you suggest, but I think the pattern would remain the same, just under a different name.

alxhub replied:
Yes, that's what I'm suggesting - an `Element` class that contains information about the element. It'll look very similar to what you already have.

Then:

* `IndexedComponent` will have an `Element[]` array.
* We get rid of `IndexedComponent.identifiers` in favor of fields on the `Element` objects, since each `TemplateIdentifier` is logically owned by an `Element`.
* `Element` gains a `tagName` `TemplateIdentifier` which maps the actual HTML tag.

ayazhafiz replied:
Discussed in person. Resolving.

ayazhafiz commented:
Running presubmit now to make sure this introduces no breaking changes.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/src/api.ts:
```diff
@@ -33,6 +28,32 @@ export interface TemplateIdentifier {
   span: AbsoluteSourceSpan;
   kind: IdentifierKind;
 }
+export interface PropertyIdentifier extends TemplateIdentifier { kind: IdentifierKind.Property; }
```
Doc strings identifying what each of these are.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/src/template.ts:
```diff
@@ -93,11 +95,7 @@ class ExpressionVisitor extends RecursiveAstVisitor {
-      span,
-      kind,
-    });
+    this.identifiers.push({ name: ast.name, span, kind, } as TopLevelIdentifier);
```
This cast is here because the shape of the object you're creating there is not assignable to `TopLevelIdentifier`, because it won't work if `kind` happens to be `IdentifierKind.Element`.

In reality, this method isn't (and shouldn't) be called to add an `IdentifierKind.Element` identifier. So, instead of casting, let's fix the incoming type of `kind` to only those types which `addIdentifier` is called for.

ayazhafiz replied:
Moved to a different identifier type that only uses identifiers that can be created in expression visitors.

Unfortunately, still have to use a cast. This is a limitation of the TypeScript version used in Angular right now; see [this playground](https://www.typescriptlang.org/play/?ts=3.1.6#code/KYOwrgtgBA0gliAJlA3gKCpqAxA9rgGgywCEBDAJzQF800EAXYCgMzIGNgoSTVjMA1gkQAuWMKJYoLfGIDODCggDmAbhp1GzNpxz4owAB5Mkc7iXRShSMfCQA6PIQ30QTVhy7kKB46ERmPHxWwrbC9t5EtGgMAJ4ADlwAglAAvHq4AD7e6mjsuCAKUABGYkkA2gC6aVBV6ixgIOwMcAVQZBAAFNaiUBUA5D39lQCUwVjF9vFgcgAWnag9Yj0EMrgi-WT91CPq1EA) for an example. Switching to 3.5.1 resolves the issue.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/src/api.ts:
```diff
@@ -33,6 +28,32 @@ export interface TemplateIdentifier {
+ */
+export interface ElementIdentifier extends TemplateIdentifier {
+  kind: IdentifierKind.Element;
+  attributes: Set<AttributeIdentifier>;
```
Doc strings identifying what these are.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/src/template.ts:
```diff
@@ -126,8 +124,36 @@ class TemplateVisitor extends RecursiveTemplateVisitor {
+    // An element's source span can be of the form `<element>`, `<element />`, or
+    // `<element></element>`. Only the selector is interesting to the indexer, so the source is
+    // searched for the first occurrence of the element (selector) name.
+    const start = sourceSpan.start.offset + sourceSpan.toString().indexOf(name);
```
Maybe add an assertion to verify this is always found, and not `-1`.

alxhub reviewed (changes requested):
Otherwise this LGTM.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/src/template.ts:
```diff
@@ -97,7 +106,7 @@ class ExpressionVisitor extends RecursiveAstVisitor {
       span,
       kind,
-    });
+    } as Id);
```
The cast is definitely not needed if `kind` is typed correctly, even in earlier versions of TS. See [this example](https://www.typescriptlang.org/play/?ts=3.1.6#code/KYOwrgtgBA0gliAJlA3gKCpqAxA9rgGgywCEBDAJzQF800EAXYCgMzIGNgd9VjMBrBIgBcsIQDo8uANw06jZm05RyFXliiCko+EnGrZGgEaVRAZwYUEAc1m00DAJ4AHLgBUAFjagBebrgAfAzp2XBALKBZ8M1FPGwBtAF1fKCTZFjAQdgY4MKgyCAAKLRExPSkASnUsAHoaqAYvMyg4ZrIAG3bcAHdgZA9mLiNgdjIwMy4yEChcIwArEYYobrhGzSEdCSkoUZAQXCXhmZYGly5VIg06qDNcFqXEXGAzEAByJZBgPvyoV5MKV5QZwUXCuChOcR8SLRcTOcYeQqoEqiErUCp2NBAA).

ayazhafiz committed:
feat(ivy): index template elements for selectors, attributes, directives

Add support for indexing elements in the indexing module.
Opening and self-closing HTML tags have their selector indexed, as well
as the attributes on the element and the directives applied to an
element.

alxhub commented on packages/compiler-cli/src/ngtsc/indexer/src/template.ts:
```diff
@@ -71,7 +74,8 @@ class ExpressionVisitor extends RecursiveAstVisitor {
    * @param kind identifier kind
    */
-  private visitIdentifier(ast: AST&{name: string, receiver: AST}, kind: IdentifierKind) {
+  private visitIdentifier<Id extends ExpressionIdentifier>(
```
I don't think `visitIdentifier` needs to be generic. It doesn't use the generic type for anything that you couldn't name a specific type for.

alxhub reviewed (approved):
This LGTM other than the generic - please fix before submitting!

ayazhafiz committed:
fixup! feat(ivy): index template elements for selectors, attributes, directives
